File: java/problem282.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package de.pedemo;

import java.math.BigInteger;

/**
 *
 * @author antonstepan
 */
public class peDemo {

    public static void main(String[] args) {
		
		System.out.println("(1:1)" + acklow(1,1));
		System.out.println("(2:2)" + acklow(2,2));
		System.out.println("(3:3)" + acklow(3,3));
		System.out.println("(4:4)" + ack(new BigInteger("4"), new BigInteger("4")));
	}

	public static BigInteger hyper(BigInteger a, BigInteger n, BigInteger b) {
		if (n.compareTo(new BigInteger("1")) == 0) return a.modPow(b, new BigInteger("14").pow(8));
		if (b.compareTo(new BigInteger("0")) == 0) return new BigInteger("1");
		return hyper(a, n.subtract(new BigInteger("1")), hyper(a,n,b.subtract(new BigInteger("1"))));
	}

	public static BigInteger ack(BigInteger m, BigInteger n) {

		return hyper(new BigInteger("2"), m.subtract(new BigInteger("2")), (n.add(new BigInteger("3")))).subtract(new BigInteger("3"));
	}
	
	public static int acklow(int m, int n) {
		if (m==0) return n+1;
		if (n==0) return acklow(m-1,1);
		return acklow(m-1,acklow(m,n-1));
	}
}

File: python/problem282.py
# Anton Stepan
# Problem282 - projecteuler.com
# ackermann shit

def hyper(a,n,b):
  if (n==1): return pow(a,b,14**8)
  if (b==0): return 1
  return hyper(a,n-1, hyper(a,n,b-1))

def ack(m,n):
  #if (n<4): return acklow(m,n)
  return hyper(2, m-2, n+3) - 3

def acklow(m,n):
  if (m==0): return n+1
  if (n==0): return acklow(m-1,1)
  return acklow(m-1, acklow(m,n-1))

#print '1,1 = ' + str(ack(1,1))
#print '2,2 = ' + str(ack(2,2))
#print '3,3 = ' + str(ack(3,3))
#print '4,4 = ' + str(ack(4,4))
print '5,5 = ' + str(ack(4,1))
print '5,5 = ' + str(ack(4,2))
print '5,5 = ' + str(ack(4,3))
print '5,5 = ' + str(ack(4,4))
print '5,5 = ' + str(ack(4,5))
print '5,5 = ' + str(ack(5,2))